lib/pacemaker/pcmk_verify.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2023 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+#include <crm/cib/internal.h>
+#include <crm/common/output.h>
+#include <crm/common/results.h>
+#include <crm/common/scheduler.h>
+#include <pacemaker-internal.h>
+#include <pacemaker.h>
+
+#include <stdint.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "libpacemaker_private.h"
+
+int
+pcmk__parse_cib(pcmk__output_t *out, const char *cib_source, xmlNodePtr *cib_object)
+{
+    int rc = pcmk_rc_ok;
+
+    crm_info("=#=#=#=#= Getting XML =#=#=#=#=");
+
+    if (cib_source == NULL) {
+        crm_info("Reading XML from: live cluster");
+        rc = cib__signon_query(out, NULL, cib_object);
+
+    }  else if (cib_source[0] == '<') {
+        *cib_object = string2xml(cib_source);
+        if (*cib_object == NULL) {
+            rc = ENODATA;
+        }
+
+    } else if (cib_source[0] == '-') {
+        *cib_object = stdin2xml();
+        if (*cib_object == NULL) {
+            rc = ENODATA;
+        }
+
+    } else {
+        *cib_object = filename2xml(cib_source);
+        if (*cib_object == NULL) {
+            rc = ENODATA;
+        }
+    }
+
+    if (rc == ENODATA) {
+        out->err(out, "Couldn't parse input file");
+    }
+
+    return rc;
+}
+
+int
+pcmk__verify(pcmk_scheduler_t *scheduler, pcmk__output_t *out, xmlNode *cib_object)
+{
+    int rc = pcmk_rc_ok;
+    xmlNode *status = NULL;
+
+    if (!pcmk__xe_is(cib_object, XML_TAG_CIB)) {
+        rc = EBADMSG;
+        out->err(out, "This tool can only check complete configurations (i.e. those starting with <cib>).");
+        goto verify_done;
+    }
+
+    status = pcmk_find_cib_element(cib_object, XML_CIB_TAG_STATUS);
+    if (status == NULL) {
+        create_xml_node(cib_object, XML_CIB_TAG_STATUS);
+    }
+
+    if (pcmk__validate_xml(cib_object, NULL, (xmlRelaxNGValidityErrorFunc) out->err, out) == FALSE) {
+        crm_config_error = TRUE;
+        rc = pcmk_rc_schema_validation;
+        free_xml(cib_object);
+        cib_object = NULL;
```

Comment:
The caller should handle freeing the XML object (if we wanted to handle it here, the argument would have to be `xmlNode **` to NULL it for the caller, but the caller might want to do something else with the XML, so it's better to let the caller handle it)

Reply:
If `pcmk__schedule_actions()` is called, then a later call to `pe_free_working_set()` frees the XML object. So if we want the caller to free the XML object, we need to know whether or not `pcmk__verify()` calls `pcmk__schedule_actions()`. This would mean e.g. introducing a variable that stores this information. Do you think this is a good idea, or do you think it'd be cleaner to not have the caller free the XML object?